### PR TITLE
feat: Add pipeline evaluation module

### PR DESCRIPTION
**Pull request by dataJSA**
This PR introduce a dedicated module for streamlining evaluation of our scoring pipelines.
 
 The new functions will allow to persist model performance : 
 - Add performance diagnosis plots
 - Add classification reports (csv,  json) 
 
 Here is how it looks for now:
 
image
 
Let me know what you think!

### CODE REVIEW

File: pyro_risks/evaluate.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    parser.add_argument(
+        "--destination", default=None, help="Folder where the report should be saved."
+    )
+    return parser.parse_args(args)
+
+
+if __name__ == "__main__":
+
+    args = parse_args(sys.argv[1:])
+    main(args)
```
Comment (frgfm) [inline pyro_risks/evaluate.py:203]: Just my point of view but: if this is a library, scripts should be clearly separated from modules. We would have to move train.py as well, but I suggest to keep everything in the `pyro_risks` folder as a module and not a script

Comment (frgfm) [reply pyro_risks/evaluate.py:203]: (You could add the script part in the `scripts` folder)

**Review by frgfm — COMMENTED** [review]
Thanks for the PR! Very nice plots, I just added a suggestion but I'm not the best reviewer for these features

File: pyro_risks/evaluate.py (hunk 1 above)
Comment (dataJSA) [reply pyro_risks/evaluate.py:203]: Thanks for the review! I understand what you are proposing but I also think that having entry points directly in the package is valuable. 

A middle ground could be to refactor all the scripts into a single entry and to built a CLI with nested command. We could then install the package from GitHub or PyPi and directly run: 
 
```bash
pyrorisks evaluate --destination ./metadata-registry
pyrorisks train --destination ./model-registry
```

The package tree could look like that which is better than what I am doing now for sure 

```bash
├── config
│   ├── __init__.py
│   ├── datasets.py
│   ├── logger.py
│   └── models.py
├── models
│   ├── __init__.py
│   ├── pipelines.py
│   ├── transformers.py
│   └── utils.py
├── pipelines
│   ├── load.py
│   ├── evaluate.py
│   ├── predict.py
│   ├── train.py
├──  main.py
├──  version.py
```


I will open an issue for that ! 

In the meantime these entry points at the end of the files are really helpful and helps with trying out machine learning orchestration frameworks as mentioned in the issue #37 and I have proposals that I want to discuss soon with the team in another issue 😄 

Let me know if It sounds right to you !

File: pyro_risks/evaluate.py [hunk 2]
```diff
@@ -157,14 +158,7 @@ def predict(x):
...
 def main(args):
-    usecols = [cfg.DATE_VAR, cfg.ZONE_VAR, cfg.TARGET] + cfg.PIPELINE_ERA5T_VARS
-    pipeline_vars = [cfg.DATE_VAR, cfg.ZONE_VAR] + cfg.PIPELINE_ERA5T_VARS
-    df = pd.read_csv(cfg.ERA5T_VIIRS_PIPELINE, usecols=usecols)
-    df["day"] = df["day"].apply(
-        lambda x: datetime.strptime(str(x), "%Y-%m-%d") if not pd.isnull(x) else x
-    )
-    X = df[pipeline_vars]
-    y = df[cfg.TARGET]
+    X, y = load_dataset()
```
Comment (frgfm) [inline pyro_risks/evaluate.py:161]: The refactoring is nice, but I'm not sure about the point here?
You removed the script part but kept "def main...". As you suggested in the previous comments, either we keep those and accept scripts inside the lib, or we move everything outside to another folder. Either way, I wouldn't recommend keeping `def main` if we aren't turning this into a script :sweat_smile:

File: pyro_risks/load.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    Load Pyro Risks training datasets.
+
+    Download and load Pyro Risks training datasets.
+
+    Returns:
+        Tuple[pd.DataFrame, pd.Series]: [description]
+    """
+    dataset_path = os.path.join(cfg.DATA_REGISTRY, cfg.DATASET)
+    usecols = [cfg.DATE_VAR, cfg.ZONE_VAR, cfg.TARGET] + cfg.PIPELINE_ERA5T_VARS
+    pipeline_vars = [cfg.DATE_VAR, cfg.ZONE_VAR] + cfg.PIPELINE_ERA5T_VARS
```
Comment (frgfm) [inline pyro_risks/load.py:28]: Those could be passed as function arguments

File: pyro_risks/load.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            destination=cfg.DATA_REGISTRY,
+        )
+
+    df = pd.read_csv(dataset_path, usecols=usecols)
+    df["day"] = df["day"].apply(
+        lambda x: datetime.strptime(str(x), "%Y-%m-%d") if not pd.isnull(x) else x
+    )
+
+    X = df[pipeline_vars]
+    y = df[cfg.TARGET]
```
Comment (frgfm) [inline pyro_risks/load.py:44]: same for cfg.TARGET

File: pyro_risks/train.py [hunk 5]
```diff
@@ -126,9 +126,7 @@ def train_pipeline(
 
     elif model == "XGBOOST":
         xgb_pipeline.fit(
-            X_train,
-            y_train,
-            xgboost__eval_metric=cfg.XGB_FIT_PARAMS["eval_metric"],
+            X_train, y_train, xgboost__eval_metric=cfg.XGB_FIT_PARAMS["eval_metric"]
```
Comment (frgfm) [inline pyro_risks/train.py:129]: At some point, we'll have to decide if we go full black style or not :sweat_smile:

File: scripts/evaluate.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+# This program is licensed under the GNU Affero General Public License version 3.
+# See LICENSE or go to <https://www.gnu.org/licenses/agpl-3.0.txt> for full license details.
+
+from pyro_risks.evaluate import main, parse_args
+import sys
+
+if __name__ == "__main__":
+
+    args = parse_args(sys.argv[1:])
+    main(args)
```
Comment (frgfm) [inline scripts/evaluate.py:12]: I really think `main` should be renamed

File: scripts/load.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+from pyro_risks.datasets.utils import download
+import pyro_risks.config as cfg
+
+if __name__ == "__main__":
+    download(
+        url=cfg.ERA5T_VIIRS_PIPELINE,
+        default_extension="csv",
+        unzip=False,
+        destination=cfg.DATA_REGISTRY,
+    )
```
Comment (frgfm) [inline scripts/load.py:15]: Why do we need to have a separate script for download?

File: scripts/train.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+# This program is licensed under the GNU Affero General Public License version 3.
+# See LICENSE or go to <https://www.gnu.org/licenses/agpl-3.0.txt> for full license details.
+
+from pyro_risks.train import main, parse_args
+import sys
+
+if __name__ == "__main__":
+
+    args = parse_args(sys.argv[1:])
+    main(args)
```
Comment (frgfm) [inline scripts/train.py:12]: renaming is key :sweat_smile:

**Review by frgfm — COMMENTED** [review]
Hey there! I left many comments :sweat_smile:

File: pyro_risks/config/datasets.py [hunk 9]
```diff
@@ -22,6 +22,7 @@
 FR_FWI_2020_FALLBACK: str = f"{DATA_FALLBACK}/JRC_FWI_2020.zip"
 FR_ERA5LAND_FALLBACK: str = f"{DATA_FALLBACK}/ERA5_2019.nc"
 FR_ERA5T_FALLBACK: str = f"{DATA_FALLBACK}/era5t_2019.nc"
+DATASET: str = f"merged_era_viirs.csv"
```
[the remark this answers is not in the document]
Comment (dataJSA) [reply pyro_risks/config/datasets.py]: Yes !

File: pyro_risks/train.py (hunk 5 above)
Comment (dataJSA) [reply pyro_risks/train.py:129]: Not sure to follow,  we have added black check within the CI and they are green

File: pyro_risks/evaluate.py (hunk 2 above)
Comment (dataJSA) [reply pyro_risks/evaluate.py:161]: Ok, I had an amazing idea concerning the training prediction and orchestration and I got way to excited 😄
I wanted to explore it really quickly and also not do a large PR (but small ones) which sounds reasonable. 
However, I totally should have opened an other issue for discussing the design first!

 I will do that and I am pretty sure my code will be cleaner and make more sense and the review will be shorter for you 😆

File: pyro_risks/load.py (hunk 3 above)
Comment (dataJSA) [reply pyro_risks/load.py:28]: Yes !

File: scripts/load.py (hunk 7 above)
Comment (dataJSA) [reply scripts/load.py:15]: https://github.com/pyronear/pyro-risks/pull/50#discussion_r595346976

**Comment by frgfm** [comment]
@dataJSA not sure if it was on purpose, but you didn't delete the remaining branch ;)

**Pull request closed without merging**